hera_cal/utils.py

```diff
@@ -1251,7 +1252,9 @@ def red_average(data, reds=None, bl_tol=1.0, inplace=False,
             navg = np.sum(n * f, axis=0)         # this is the new total nsample (without flagged elements)
             fmax = np.max(f, axis=2)             # collapse along freq: marks any fully flagged integrations
             iavg = np.sum(tint.squeeze() * fmax, axis=0) / np.sum(fmax, axis=0).clip(1e-10, np.inf)
-            favg = np.isclose(wsum, 0.0)         # this is getting any fully flagged pixels
+            # average flags should be where weights x flags x nsamples sum to zero.
+            wfsum = np.sum(n * w * f, axis=0).clip(1e-10, np.inf)
```

Comment:
this seems inconsistent--`favg` should be where `wsum==0`. If someone wants the weights to be nsample * ~flag (by default) then this is what `wsum` is equal to. In the case where the wgts are directly fed, the docstring says that this supercedes nsample and flags, in which case favg (as is written) is inconsistent with what was used to actually weight the data. what is the rationale for changing this?

Reply:
The rational here is to preserve flags even when the weights are not zero where the data is flagged. Without this change, a user could provide `w` that is not zero where the data is flagged and this will effectively override the previous flags.

Reply:
Because `navg` is not determined by weights, even when the user provides them, I find the most logical interpretation of the docstring to mean that user provided weights supercede `nsamples` and `flags` as weights but do not replace `nsamples` and `flags`.

Reply:
At any rate, I do totally understand the concern for tweaks that change the API behavior (especially for `H1C`).

 Let me know if I should keep this change (and update the docstring?) or whether you'd like this change in behavior to be optional (so as not too interfere with `H1C`).

Reply:
Its not an H1C API issue for me. I actually think that it is inconsistent with the averaging: it seems that `favg` should be consistent with the averaging actually done to the data. If you pass a `wgts` dictionary, you are effectively ignoring the flags, so the "flags" in your weighting are set by when `np.isclose(wgts, 0)`. I understand that is it weird to ignore the flags, but that is the spirit of the functionality enabled by the `wgts` dictionary. If we want to remove that functionality that is a separate question, but I think `favg` should be consistent with what we weight by. Lmk if that makes sense.

Yeah the nsamples is an outlier, that might be changed to
`navg = np.sum(n * ~np.isclose(w, 0),  axis=0)`

Reply:
I think I see where Nick is coming from here. 

Perhaps the compromise here is to make the function error if wgts is provided and flags or nsamples is also provided. And maybe the docstring can be updated for clarity.

Reply:
A mode of use that I do want to support is for a user to redundantly average data, not set that data to zero, but still have flags in those integrations.

Reply:
having weights override flags makes this impossible. The way my update is written still flags regions where the weights are zero, it simply preserves the flags that are input. If someone wants flags to only occur where weights are zero, they can provide flags that are False and weights that are zero where they wish for the flags to occur. 

I don't see though how we can allow for redundantly averaged data that is not set to zero in flagged regions using the method as written though. Perhaps a `flagging_mode` keyword can be used to fix this.

Reply:
A way that this could be done would be to perform the averaging with the user provided weights and then restore the flags afterwards. I'm open to writing a wrapper for this.

Reply:
Well, the whole idea of `wgts` is to allow the user to do what they want. If they want to use the flags but have interesting, non-standard wgts, they can simply construct that themselves (`wgts = my_wgt * ~flgs`)--that's the point of a user-fed option.

Reply:
If one does this, then they will preserve the flags, but the data will now be zero in the flagged channels so by choosing to keep the flags, the user has been forced to throw away the data in those flagged channels.

Reply:
I think that goes against the core philosophy of what a flag represents though. If someone wants to not weight by the flags, that's exactly what `wgts` is trying to enable--weights set by the user. At the same time, if one wants to keep the initial set of flags, they can keep them. nothing is stopping them, right? But, for the function itself to report that a pixel is "flagged" (via `favg`) but is actually not set to zero (or vice versa for that matter) is inconsistent. I mean, I think what you are asking for is perhaps achievable by making _another_ array, say `avg_inp_flags`, which takes the average of the input flags, regardless of what `wgts` is, however, this should not replace `favg` in my opinion.

Reply:
I agree that it's kind of weird philosophically but I do think that it's the right kind of redundant averaging for us to do until we are super comfortable with our flags. I toyed around with the wrapper approach and things get kind of messy (you have to compute redundant baselines and filter them within the wrapper) so I think the best way forward is to add an argument to `red_average` that is defaulted to the case where we have weights determine flags and compute nsamples as you suggest: `navg = np.sum(n * ~np.isclose(w, 0), axis=0)`. In the non-default case it will compute `favg` as the product of `n`, `w`, and `f`.